subsys/net/lib/websocket/websocket.c

```diff
@@ -0,0 +1,485 @@
+/*
+ * Copyright (c) 2017 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_NET_DEBUG_WEBSOCKET)
+#define SYS_LOG_DOMAIN "ws"
+#define NET_SYS_LOG_LEVEL SYS_LOG_LEVEL_DEBUG
+#define NET_LOG_ENABLED 1
+#endif
+
+#include <zephyr.h>
+#include <string.h>
+#include <strings.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <version.h>
+
+#include <net/net_ip.h>
+#include <net/websocket.h>
+
+#include <mbedtls/base64.h>
+#include <mbedtls/sha1.h>
+
+#define BUF_ALLOC_TIMEOUT 100
+
+#define HTTP_CRLF "\r\n"
+
+/* From RFC 6455 chapter 4.2.2 */
+#define WS_MAGIC "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
+
+static void ws_mask_payload(u8_t *payload, size_t payload_len,
+			    u32_t masking_value)
+{
+	int i;
+
+	for (i = 0; i < payload_len; i++) {
```

Comment:
There should be `websocket_ctx` object, what is "i" in this function, should be part of this context.

Reply:
There is no need for separate websocket_ctx, all context info is handled currently by http_ctx.

Reply:
Couple of points:

1. For a server implementation (e.g. HTTP server), there should be a context object. However, for each request it serves, there should be a separate *request context* object. The need for that should be obvious, considering a server may serve several independent requests (== clients). I had a feeling that https://github.com/zephyrproject-rtos/zephyr/pull/4243 doesn't have that, didn't raise it before to not procrastinate it further, but without a proper implementation of the above, it's an alpha version.

2. Websocket is a separate protocol from HTTP, so need a separate WS request (connection) context from HTTP request ctx. Implementation-wise, those can be the same underlying structure (i.e. if WS support enabled, every "mere HTTP" req ctx gets overhead of WS fields).